### PR TITLE
Migrate Generally Useful Objects From Alerting (Monitor, Execute, Index)

### PR DESCRIPTION
**Pull request by okram**
### Description
Alerting has various POJOs that are useful beyond the boundaries of Alerting. In particular, for this push, Monitor was isolated along with those classes within its referential reach. Furthermore, the actions ExecuteMonitor and IndexMontitor (i.e. Create) were migrated as well.
 
### Issues Resolved
Part 1 in the multipart/multi-repository Security Analytics Plugin Issue #10.

### Check List
- [X] New functionality includes testing.
  - [X] All tests pass
- [O] New functionality has been documented.
  - [X] New functionality has javadoc added
- [X] Commits are signed per the DCO using --signoff 

Integration testing was done on MacOSX using a Linked Docker environment.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Review by getsaurabh02 — CHANGES_REQUESTED** [review]
Thanks @okram for taking a shot on this. While introducing these abstraction for plugin to plugin transport communications lets keep below considerations:

1. Common-utils should only have the (models) PoJos and Action Interfaces required for Transport interactions. 
2.  All execution/business logic including the core artifacts (such as Settings) should continue to be owned by the Alerting Plugin itself.
3. To decouple the object and code dependency - we can take the liberty of introducing new model classes in common-utils and use them in the heavy weight (monolithic) classes in Alerting. Such as Monitor object. Instead of moving the monolithic class directly. 

Since we will need to work towards the right abstraction, do you think can we wait for right it to come in soon - even before we merge this PR into the dev branch? 

In case you think if it's going to take some time and its absolutely necessary to unblock parallel stream of work (in Alerting and SecurityAnalytics), we can do a temp merge in dev branch with a note that this should NOT be merged in mainline.

File: src/main/kotlin/org/opensearch/commons/alerting/resthandler/RestExecuteMonitorAction.kt [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+import org.opensearch.rest.BaseRestHandler
+import org.opensearch.rest.RestHandler
+import org.opensearch.rest.RestRequest
+import org.opensearch.common.xcontent.XContentParser.Token.START_OBJECT
+import org.opensearch.commons.utils.logger
+import org.opensearch.rest.RestRequest.Method.POST
+import org.opensearch.rest.action.RestToXContentListener
+import java.time.Instant
+
+class RestExecuteMonitorAction : BaseRestHandler() {
```
[the remark this answers is not in the document]
Comment (okram) [reply src/main/kotlin/org/opensearch/commons/alerting/resthandler/RestExecuteMonitorAction.kt:18]: Okay. I'll put those back in Alerting.

File: src/main/kotlin/org/opensearch/commons/OpenSearchExtensions.kt [hunk 2]
```diff
@@ -0,0 +1,257 @@
...
+fun ToXContent.convertToMap(): Map<String, Any> {
+    val bytesReference = XContentHelper.toXContent(this, XContentType.JSON, false)
+    return XContentHelper.convertToMap(bytesReference, false, XContentType.JSON).v2()
+}
+
+/**
+ * Backs off and retries a lambda that makes a request. This should not be called on any of the [standard][ThreadPool]
+ * executors since those executors are not meant to be blocked by sleeping.
+ */
+fun <T> BackoffPolicy.retry(block: () -> T): T {
```
[the remark this answers is not in the document]
Comment (okram) [reply src/main/kotlin/org/opensearch/commons/OpenSearchExtensions.kt:55]: So `OpenSearchExtensions.kt` is one of those Kotlin "scripts." I can gut the methods needed and put them with their object and then simply get rid of all the Kotlin "scripts".

File: src/main/kotlin/org/opensearch/commons/alerting/model/LegacyOpenDistroAlertingSettings.kt [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+
+import org.opensearch.common.settings.Setting
+import org.opensearch.common.unit.TimeValue
+import java.util.concurrent.TimeUnit
+
+/**
+ * Legacy Opendistro settings specific to [AlertingPlugin]. These settings include things like history index max age, request timeout, etc...
+ */
+
+class LegacyOpenDistroAlertingSettings {
```
[the remark this answers is not in the document]
Comment (okram) [reply src/main/kotlin/org/opensearch/commons/alerting/model/LegacyOpenDistroAlertingSettings.kt:16]: Okay. These settings are refered to in the various classes migrated. How do you recommend these settings be specified?

File: src/main/kotlin/org/opensearch/commons/alerting/model/Monitor.kt [hunk 4]
```diff
@@ -0,0 +1,323 @@
...
+    override val lastUpdateTime: Instant,
+    override val enabledTime: Instant?,
+    // TODO: Check how this behaves during rolling upgrade/multi-version cluster
+    //  Can read/write and parsing break if it's done from an old -> new version of the plugin?
+    val monitorType: MonitorType,
+    val user: User?,
+    val schemaVersion: Int = NO_SCHEMA_VERSION,
+    val inputs: List<Input>,
+    val triggers: List<Trigger>,
+    val uiMetadata: Map<String, Any>
```
[the remark this answers is not in the document]
Comment (okram) [reply src/main/kotlin/org/opensearch/commons/alerting/model/Monitor.kt:50]: That' be great if there was a simpler `Monitor`. Most of the code pulled over is becauuse `Monitor` is so beefy. If you can kill some of those fields and still get the functionality you are looking for, that'd be golden.

File: src/main/kotlin/org/opensearch/commons/alerting/resthandler/RestExecuteMonitorAction.kt (hunk 1 above)
[the remark this answers is not in the document]
Comment (okram) [reply src/main/kotlin/org/opensearch/commons/alerting/resthandler/RestExecuteMonitorAction.kt]: Latest push to this PR has the REST handlers removed.

File: src/main/kotlin/org/opensearch/commons/OpenSearchExtensions.kt (hunk 2 above)
[the remark this answers is not in the document]
Comment (okram) [reply src/main/kotlin/org/opensearch/commons/OpenSearchExtensions.kt]: By removing classes, I was able to remove this method (along with others). I've gutted all unused methods in the Kotlin "scripts" (meta-programming extensions).

File: src/main/kotlin/org/opensearch/commons/alerting/JobScheduler.kt [hunk 5]
```diff
@@ -0,0 +1,228 @@
...
+import java.util.concurrent.ConcurrentHashMap
+import java.util.concurrent.TimeUnit
+import java.util.stream.Collectors
+
+/**
+ * JobScheduler is a class for scheduling and descheduling ScheduleJobs. This class keeps list of ScheduledJob Ids that are currently scheduled.
+ *
+ * JobScheduler is unaware of the ScheduledJob version and it is up to callers to ensure that the older version of ScheduledJob to be descheduled and schedule the new version.
+ */
+class JobScheduler(private val threadPool: ThreadPool, private val jobRunner: JobRunner) {
```
[the remark this answers is not in the document]
Comment (okram) [reply src/main/kotlin/org/opensearch/commons/alerting/JobScheduler.kt]: Was able to remove `JobRunner`, `JobScheduler`, and `JobSchedulerMetrics`.

File: src/main/kotlin/org/opensearch/commons/alerting/model/Action.kt [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+data class Action(
+    val name: String,
+    val destinationId: String,
+    val subjectTemplate: Script?,
+    val messageTemplate: Script,
+    val throttleEnabled: Boolean,
+    val throttle: Throttle?,
+    val id: String = UUIDs.base64UUID(),
+    val actionExecutionPolicy: ActionExecutionPolicy? = null
+) : Writeable, ToXContentObject {
```
[the remark this answers is not in the document]
Comment (okram) [reply src/main/kotlin/org/opensearch/commons/alerting/model/Action.kt:32]: I was unable remove `Action`. It's fairly tangled in `Monitor`. Will now do as you recommend with a "mini-Action" implementation.

File: src/main/kotlin/org/opensearch/commons/alerting/model/Monitor.kt (hunk 4 above)
[the remark this answers is not in the document]
Comment (okram) [reply src/main/kotlin/org/opensearch/commons/alerting/model/Monitor.kt:51]: What are the attributes you believe should be isolated? Only primitives (no objects)?  If so, that would create a cascade of deletes: `Trigger`, `TriggerRunResult`, `Input`, `BuckerLevelTrigger`, `BucketSelector*` (like 5 classes).

If having `Monitor` only contain primitives is legitimate, then tell me and I'll put my hazmat suit back on, pull out my chainsaw, and start butchering away!  <with a crazed grin on my face>

File: src/main/kotlin/org/opensearch/commons/alerting/model/AlertIndices.kt [hunk 7]
```diff
@@ -0,0 +1,555 @@
...
+ * in [ALERT_INDEX].  Completed alerts are written to [ALERT_HISTORY_WRITE_INDEX] which is an alias that points at the
+ * current index to which completed alerts are written. [ALERT_HISTORY_WRITE_INDEX] is periodically rolled over to a new
+ * date based index. The frequency of rolling over indices is controlled by the `opendistro.alerting.alert_rollover_period` setting.
+ *
+ * These indexes are created when first used and are then rolled over every `alert_rollover_period`. The rollover is
+ * initiated on the cluster manager node to ensure only a single node tries to roll it over.  Once we have a curator functionality
+ * in Scheduled Jobs we can migrate to using that to rollover the index.
+ */
+// TODO: reafactor to make a generic version of this class for finding and alerts
+class AlertIndices(
```
[the remark this answers is not in the document]
Comment (okram) [reply src/main/kotlin/org/opensearch/commons/alerting/model/AlertIndices.kt]: This has been removed.

File: src/main/kotlin/org/opensearch/commons/alerting/model/LegacyOpenDistroAlertingSettings.kt (hunk 3 above)
[the remark this answers is not in the document]
Comment (okram) [reply src/main/kotlin/org/opensearch/commons/alerting/model/LegacyOpenDistroAlertingSettings.kt]: There were only a few `DEFAULT` settings that were ultimately referenced in this class. I moved them to `const` in the the respective depending classes. Thus, this has been removed.

**Comment by lezzago** [comment]
The approach has changed, so closing this PR

**Pull request closed without merging**